runtime/doc/lua-plugin.txt

```diff
@@ -0,0 +1,432 @@
+Lazy loading					      *lua-plugin-lazy-loading*
+
+When it comes to initializing your plugin, assume your users may not be using
+a plugin manager that takes care of lazy loading for you.
+Making sure your plugin does not unnecessarily impact startup time is your
+responsibility. A plugin's functionality may evolve over time, potentially
+leading to breakage if users have to hack into the loading mechanisms.
+Furthermore, a plugin that implements its own lazy initialization properly will
+likely have less overhead than the mechanisms used by a plugin manager or user
+to load that plugin lazily.
+
+------------------------------------------------------------------------------
+Defer `require` calls			*lua-plugin-lazy-loading-defer-require*
+
+|plugin| scripts should not eagerly `require` Lua modules.
+
+For example, instead of:
+>lua
+    local foo = require("foo")
+    vim.api.nvim_create_user_command("MyCommand", function()
+        foo.do_something()
+    end, {
+      -- ...
+    })
+<
+which will eagerly load the `foo` module and any other modules it imports
+eagerly, you can lazy load it by moving the `require` into the command's
+implementation.
+>lua
+    vim.api.nvim_create_user_command("MyCommand", function()
+        local foo = require("foo")
+        foo.do_something()
+    end, {
+      -- ...
+    })
+<
+NOTE: For a Vimscript alternative to `require`, see |autoload|.
+
+NOTE: In case you are worried about eagerly creating user commands, autocommands
+      or keymaps at startup:
+      Plugin managers that provide abstractions for lazy-loading plugins on
+      such events will need to create these themselves.
+
+------------------------------------------------------------------------------
+Filetype-specific functionality		     *lua-plugin-lazy-loading-filetype*
+
+Consider making use of |filetype| for any functionality that is specific to a
+filetype, by putting the initialization logic in a `ftplugin/{filetype}.lua`
+script.
+
+------------------------------------------------------------------------------
+Example				     *lua-plugin-lazy-loading-filetype-example*
+
+A plugin tailored to Rust development might have initialization in 
+`ftplugin/rust.lua`:
+>lua
+    if not vim.g.loaded_my_rust_plugin then
+        -- Initialise
+    end
+    -- NOTE: Using `vim.g.loaded_` prevents the plugin from initializing twice
+    -- and allows users to prevent plugins from loading
+    -- (in both Lua and Vimscript).
+    vim.g.loaded_my_rust_plugin = true
+
+    local bufnr = vim.api.nvim_get_current_buf()
+    -- do something specific to this buffer,
+    -- e.g. add a |<Plug>| mapping or create a command
+    vim.keymap.set("n", "<Plug>(MyPluginBufferAction)", function() 
+        print("Hello")
+    end, { noremap = true, buffer = bufnr, })
+<
+==============================================================================
+Configuration				             *lua-plugin-configuration*
```

Comment:
This whole section feels out of place here. If there's some part of the validation functionality that is missing from `vim.validate`, then we should try to improve the functionality in core rather than have every plugin author implement/augment validation functionality on their own.